oauth2/consent_strategy_extra.go

```diff
@@ -0,0 +1,163 @@
+		if _, ok := t.Method.(*jwt.SigningMethodRSA); !ok {
+			return nil, errors.Errorf("Unexpected signing method: %v", t.Header["alg"])
+		}
+
+		pk, err := s.KeyManager.GetKey(ConsentEndpointKey, "public")
+		if err != nil {
+			return nil, err
+		}
+
+		rsaKey, ok := jwk.First(pk.Keys).Key.(*rsa.PublicKey)
+		if !ok {
+			return nil, errors.New("Could not convert to RSA Private Key")
+		}
+		return rsaKey, nil
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "The consent response is not a valid JSON Web Token")
+	}
+
+	// make sure to use MapClaims since that is the default..
+	jwtClaims, ok := t.Claims.(jwt.MapClaims)
+	if err != nil || !ok {
+		return nil, errors.Errorf("Couldn't parse token: %v", err)
+	} else if !t.Valid {
+		return nil, errors.Errorf("Token is invalid")
+	}
+
+	if j, ok := session.Values["consent_jti"]; !ok {
+		return nil, errors.Errorf("Session cookie is missing anti-replay token")
+	} else if js, ok := j.(string); !ok {
+		return nil, errors.Errorf("Session cookie anti-replay value is not a string")
+	} else if js != ejwt.ToString(jwtClaims["jti"]) {
+		return nil, errors.Errorf("Session cookie anti-replay value does not match value from consent response")
+	}
+	delete(session.Values, "jti")
+
+	if time.Now().After(ejwt.ToTime(jwtClaims["exp"])) {
+		return nil, errors.Errorf("Token expired")
+	}
+
+	if ejwt.ToString(jwtClaims["aud"]) != a.GetClient().GetID() {
+		return nil, errors.Errorf("Audience mismatch")
+	}
+
+	subject := ejwt.ToString(jwtClaims["sub"])
+	if subject == "" {
+		return nil, errors.Errorf("Subject key is empty or undefined in consent response, check your payload.")
+	}
+
+	scopes := toStringSlice(jwtClaims["scp"])
+	for _, scope := range scopes {
+		a.GrantScope(scope)
+	}
+
+	var idExt map[string]interface{}
+	var atExt map[string]interface{}
+	if ext, ok := jwtClaims["id_ext"].(map[string]interface{}); ok {
+		idExt = ext
+	}
+	if ext, ok := jwtClaims["at_ext"].(map[string]interface{}); ok {
+		atExt = ext
+	}
+
+	// add key id to session headers
+	extHeader := map[string]interface{}{
+		"kid": "public",
+	}
+
+	sess := &Session{
+		DefaultSession: &openid.DefaultSession{
+			Claims: &ejwt.IDTokenClaims{
+				Audience:  a.GetClient().GetID(),
+				Subject:   subject,
+				Issuer:    s.Issuer,
+				IssuedAt:  time.Now(),
+				ExpiresAt: time.Now().Add(s.DefaultIDTokenLifespan),
+				Extra:     idExt,
+			},
+			Headers: &ejwt.Headers{extHeader},
+			Subject: subject,
+		},
+		Extra: atExt,
+	}
+
+	return sess, err
+}
+
+func (s *ConsentStrategyExtraParameters) IssueChallenge(authorizeRequest fosite.AuthorizeRequester, redirectURL string, session *sessions.Session) (string, error) {
+	token := jwt.New(jwt.SigningMethodRS256)
+	jti := uuid.New()
+	redirectURLParams, err := url.ParseQuery(redirectURL)
+	if err != nil {
+		return "", errors.WithStack(err)
+	}
+	var atExtDecoded interface{}
+	err = json.Unmarshal([]byte(redirectURLParams["at_ext"][0]), &atExtDecoded)
```

Comment:
May we have simple "tid" (or smth. similar) for tenant's srn? at_ext it is access token extra parameters and it may not correspond to url parameters

Reply:
since it's public repo I wanted to make it as generic as possible